---
title: "Deep Transcriptomic Atlas of Substantia Nigra in Parkinson's Disease"
author: "Viola Volpato"
date: 'Last update: `r date()`'
output:
  html_document:
    fig_height: 5
    fig_width: 5
    theme: cerulean
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
      toc_depth: 5
  pdf_document:
    toc: yes
  word_document:
    toc: yes
---


```{r clean r session, include=FALSE}
rm(list = ls())
```

### Install and load library

```{r load lib, echo=TRUE}
library(ggplot2)
library(Seurat)
library(RColorBrewer)
```


### Load data

```{r get raw data, echo=TRUE}
load("/nfs/dri/02/rdscw/shared/webber/SNatlas_icell8_Novaseq/manuscript_data/sn_atlas_processed.RData")

sn_atlas_processed <- NormalizeData(sn_atlas_processed)
sn_atlas_processed <- FindVariableFeatures(sn_atlas_processed, selection.method = "vst", nfeatures = 2000)
sn_atlas_processed <- ScaleData(sn_atlas_processed)
sn_atlas_processed <- RunPCA(sn_atlas_processed, features = VariableFeatures(object = sn_atlas_processed))
sn_atlas_processed <- FindNeighbors(sn_atlas_processed, dims = 1:10)
sn_atlas_processed <- FindClusters(sn_atlas_processed, resolution = 0.6)
sn_atlas_processed <- RunUMAP(sn_atlas_processed, dims = 1:10)

```

### Find cell type specific gene markers (Figure 2)

```{r gene markers, echo=TRUE}

# Method 1 (used in the manuscript):
# to be done for each cell type and subtype in controls

# level 1:

list_w_stat<-c()
df_w_test<-c()
signed_rank = function(x) sign(x) * rank(abs(x))
metadataALL<-as.data.frame(sn_atlas_processed@meta.data)
metadataALL_CTR<-metadataALL[metadataALL$Disease=="CTR",]
data<-as.matrix(GetAssayData(sn_atlas_processed, slot = "counts"))
data<-data[,colnames(data)%in%rownames(metadataALL_CTR)]
list_cell <- as.array(unique(metadataALL_CTR$CellType))
list_gene <- row.names(data)

compute_w_val<-function(gene_ref,cell_ref,matrix_expr,info_cell) {
    n_cell<- ncol(data)
    X<- matrix(-1, nrow = n_cell, ncol = 1)
    X[which(info_cell$CellSubType==cell_ref),1]<-1
    Y=t(data[gene_ref,])
    df<-data.frame(y=t(Y),x=X)
    colnames(df)<-c("y","x")
    linearMod <- lm(signed_rank(y) ~ x, data=df)
    val<-summary(linearMod)
    list_val<-val$coefficients[2,3]
    return(val$coefficients[2,3]) ## w-statisc value
}

for (i in c(1:length(list_cell))) {
    cell_ref=list_cell[i]
    print("W test")
    print(list_cell[i])
    list_w_stat<-sapply(list_gene,compute_w_val,cell_ref=list_cell[i],matrix_expr=data,info_cell=metadataALL_CTR)
    if(!exists("df_w_test")) {
        df_w_test<-data.frame(list_w_stat)
    } else {
        df_w_test<-cbind(df_w_test,list_w_stat)
    }
}
colnames(df_w_test) <- list_cell
write.table(df_w_test,"Wtest_level1",quote=F,sep="\t")

# level 2:

list_w_stat<-c()
df_w_test<-c()
signed_rank = function(x) sign(x) * rank(abs(x))
metadataALL<-as.data.frame(sn_atlas_processed@meta.data)
metadataALL_CTR<-metadataALL[metadataALL$Disease=="CTR",]
metadataALL_CTR<-metadataALL_CTR[metadataALL_CTR$CellType=="ODC",]
data<-as.matrix(GetAssayData(sn_atlas_processed, slot = "counts"))
data<-data[,colnames(data)%in%rownames(metadataALL_CTR)]
list_cell <- as.array(unique(metadataALL_CTR$CellSubType))
list_gene <- row.names(data)

compute_w_val<-function(gene_ref,cell_ref,matrix_expr,info_cell) {
    n_cell<- ncol(data)
    X<- matrix(-1, nrow = n_cell, ncol = 1)
    X[which(info_cell$CellSubType==cell_ref),1]<-1
    Y=t(data[gene_ref,])
    df<-data.frame(y=t(Y),x=X)
    colnames(df)<-c("y","x")
    linearMod <- lm(signed_rank(y) ~ x, data=df)
    val<-summary(linearMod)
    list_val<-val$coefficients[2,3]
    return(val$coefficients[2,3]) ## w-statisc value
}

for (i in c(1:length(list_cell))) {
    cell_ref=list_cell[i]
    print("W test")
    print(list_cell[i])
    list_w_stat<-sapply(list_gene,compute_w_val,cell_ref=list_cell[i],matrix_expr=data,info_cell=metadataALL_CTR)
    if(!exists("df_w_test")) {
        df_w_test<-data.frame(list_w_stat)
    } else {
        df_w_test<-cbind(df_w_test,list_w_stat)
    }
}
colnames(df_w_test) <- list_cell
write.table(df_w_test,"Wtest_ODCs",quote=F,sep="\t")



# Method 2:

Idents(sn_atlas_processed)<-sn_atlas_processed$CellType
clusterALLMarkers <- FindAllMarkers(sn_atlas_processed, only.pos = FALSE, min.pct = 0.1, thresh.use = 0.2)

```

### Prepare for MAGMA analysis (Figure 2)

```{r magma, echo=TRUE}

# level 1:

df_w_test_2p<-data.frame(cellt=c(rep("ODC",500),rep("Astrocyte",500),rep("DaN",500),rep("GABA",500),rep("OPC",500),rep("Microglia",500),rep("Tcell",500)),id=c(rownames(df_w_test[order(-df_w_test$ODC),])[1:500],rownames(df_w_test[order(-df_w_test$Astrocyte),])[1:500],rownames(df_w_test[order(-df_w_test$DaN),])[1:500],rownames(df_w_test[order(-df_w_test$GABA),])[1:500],rownames(df_w_test[order(-df_w_test$OPC),])[1:500],rownames(df_w_test[order(-df_w_test$Microglia),])[1:500],rownames(df_w_test[order(-df_w_test$Tcell),])[1:500]))

NCBI37.3.gene.loc<-read.delim("NCBI37.3.gene.loc",h=F)
names(NCBI37.3.gene.loc)[6]<-"id"
df<-merge(df_w_test_2p[,1:2],NCBI37.3.gene.loc,by=c("id"))
list_gene<-c()
for(m in unique(df$cellt)){
 val<-c(m,as.character(df[df$cellt==m,3]))
 val<-paste(val,collapse =" ")
 list_gene<-c(list_gene,val)
}
write.table(data.frame(list_gene),file="input_magma_level1_2p",quote=FALSE,row.names=FALSE,col.names=FALSE,sep="\t")

```

### Cell type-specific PPI networks (Figure 2)

```{r cellt_PPI, echo=TRUE}
library(igraph)

# to be done for DaN_3, ODC_2 and OPC (where PD genetic risk converged)

ppi_net<-read.delim("ppi_net_new2023short",h=F)
ppi_net_dan3top1000<-ppi_net[ppi_net[,1]%in%rownames(df_w_test_DaN[oder(-df_w_test_DaN$DaN_3),])[1:1000] & ppi_net[,2]%in%rownames(df_w_test_DaN[oder(-df_w_test_DaN$DaN_3),])[1:1000] ,]
ppi_net_dan3top1000_genes<-unique(c(as.character(ppi_net_dan3top1000[,1]),as.character(ppi_net_dan3top1000[,2])))

net_1 <- graph_from_data_frame(d= ppi_net_dan3top1000, vertices= ppi_net_dan3top1000_genes, directed=F)
cl_net_1<-cluster_louvain(net_1)
memb_net_1<-as.matrix(membership(cl_net_1))

memb_net_dan3top1000ppi_df<-data.frame(id=rownames(memb_net_1),module=memb_net_1)


```


### Changes in cell type proportions (Figure 3)

```{r cell_prop, echo=TRUE}

library("scProportionTest")
library(speckle)
library(limma)
library(ggplot2)

# Method 1, use FDR-corrected p-values:
prop_test <- sc_utils(sn_atlas_processed)

prop_test <- permutation_test(
	prop_test, cluster_identity = "CellSubType",
	sample_1 = "CTR", sample_2 = "PD_B5-6",
	sample_identity = "Disease"
)

# Method 2, use nominal p-values:
sn_atlas_processed_tmp<-subset(sn_atlas_processed,subset=Disease=="CTR" | Disease=="PD_B5-6")
propeller(clusters = sn_atlas_processed_tmp$CellSubType, sample = sn_atlas_processed_tmp$Sample_v2, 
          group = sn_atlas_processed_tmp$Disease)

plotCellTypeProps(clusters=sn_atlas_processed_tmp$seurat_clusters, sample=sn_atlas_processed_tmp$Disease)

sn_atlas_processed_tmp<-subset(sn_atlas_processed,subset=Disease=="CTR" | Disease=="PD_B5-6")
df_tmp<-data.frame(table(sn_atlas_processed_tmp$CellSubType))
prop_df_ctr_pd56<-data.frame(table(sn_atlas_processed_tmp$CellSubType,sn_atlas_processed_tmp$Disease))
prop_df_ctr_pd56$prop<-prop_df_ctr_pd56$Freq/df_tmp$Freq*100
pdf("celltype_prop_CTR_PD56.pdf",6,3)
ggplot(data=prop_df_ctr_pd56, aes(x=Var1, y=prop, fill=Var2)) +
     geom_bar(stat="identity") + scale_fill_manual(values=c('mediumorchid','skyblue')) + theme(axis.text.x = element_text( color = "black",size = 10, angle = 45,hjust=1))
dev.off()


props.data <- getTransformedProps(clusters=sn_atlas_processed_tmp$CellSubType, sample=sn_atlas_processed_tmp$Sample_v2,
                                  transform="logit")
df<-data.frame(props.data$Proportions)
df$disease<-ifelse(df$sample=="15_013" | df$sample=="PD_792" | df$sample=="18_005" | df$sample=="PD_714" | df$sample=="PD_816","PD56","CTR")

pdf("proportion_changes_propeller.pdf",10,3)
ggplot(df, aes(x=clusters, y=Freq, color=disease)) +
     geom_boxplot(position=position_dodge(0.8))+
     geom_jitter(position=position_dodge(0.8))+
     scale_color_manual(values=c("mediumorchid","skyblue"))+
     theme(axis.text.x = element_text( color = "black",size = 10, angle = 45,hjust=1))
dev.off()

```

### Find DEGs CTR vs PD within each cell type (Figure 3)

```{r deg_by_cellType, echo=TRUE}

# level 1:
sn_atlas_processed$clust.disease <- paste(sn_atlas_processed$CellType, sn_atlas_processed$Disease, sep = "_")
sn_atlas_processed$cluster <- Idents(sn_atlas_processed)
Idents(sn_atlas_processed) <- "clust.disease"

CTRvsPD56_DaN <- FindMarkers(sn_atlas_processed, ident.1 = "DaN_CTR", ident.2 = "DaN_PD_B5-6", verbose = FALSE)
CTRvsPD56_ODC <- FindMarkers(sn_atlas_processed, ident.1 = "ODC_CTR", ident.2 = "ODC_PD_B5-6", verbose = FALSE)
CTRvsPD56_Astrocyte <- FindMarkers(sn_atlas_processed, ident.1 = "Astrocyte_CTR", ident.2 = "Astrocyte_PD_B5-6", verbose = FALSE)
CTRvsPD56_Microglia <- FindMarkers(sn_atlas_processed, ident.1 = "Microglia_CTR", ident.2 = "Microglia_PD_B5-6", verbose = FALSE)
CTRvsPD56_OPC <- FindMarkers(sn_atlas_processed, ident.1 = "OPC_CTR", ident.2 = "OPC_PD_B5-6", verbose = FALSE)
CTRvsPD56_GABA <- FindMarkers(sn_atlas_processed, ident.1 = "GABA_CTR", ident.2 = "GABA_PD_B5-6", verbose = FALSE)
CTRvsPD56_Tcell <- FindMarkers(sn_atlas_processed, ident.1 = "Tcell_CTR", ident.2 = "Tcell_PD_B5-6", verbose = FALSE)

# level 2:
sn_atlas_processed$clust.disease <- paste(sn_atlas_processed$CellSubType, sn_atlas_processed$Disease, sep = "_")
sn_atlas_processed$cluster <- Idents(sn_atlas_processed)
Idents(sn_atlas_processed) <- "clust.disease"

opc_d56<-FindMarkers(sn_atlas_processed,ident.1 = "OPC_PD_B5-6", ident.2 = "OPC_CTR")
tcell_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Tcells_PD_B5-6", ident.2 = "Tcells_CTR")
odc2_d56<-FindMarkers(sn_atlas_processed,ident.1 = "ODC_2_PD_B5-6", ident.2 = "ODC_2_CTR")
odc1_d56<-FindMarkers(sn_atlas_processed,ident.1 = "ODC_1_PD_B5-6", ident.2 = "ODC_1_CTR")
odc0_d56<-FindMarkers(sn_atlas_processed,ident.1 = "ODC_0_PD_B5-6", ident.2 = "ODC_0_CTR")
micro2_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Microglia_2_PD_B5-6", ident.2 = "Microglia_2_CTR")
micro1_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Microglia_1_PD_B5-6", ident.2 = "Microglia_1_CTR")
micro0_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Microglia_0_PD_B5-6", ident.2 = "Microglia_0_CTR")
gaba_d56<-FindMarkers(sn_atlas_processed,ident.1 = "GABA_PD_B5-6", ident.2 = "GABA_CTR")
astro3_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Astrocyte_3_PD_B5-6", ident.2 = "Astrocyte_3_CTR")
astro2_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Astrocyte_2_PD_B5-6", ident.2 = "Astrocyte_2_CTR")
astro0_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Astrocyte_0_PD_B5-6", ident.2 = "Astrocyte_0_CTR")
astro1_d56<-FindMarkers(sn_atlas_processed,ident.1 = "Astrocyte_1_PD_B5-6", ident.2 = "Astrocyte_1_CTR")
dan2_d56<-FindMarkers(sn_atlas_processed,ident.1 = "DaN_2_PD_B5-6", ident.2 = "DaN_2_CTR")
dan1_d56<-FindMarkers(sn_atlas_processed,ident.1 = "DaN_1_PD_B5-6", ident.2 = "DaN_1_CTR")
dan0_d56<-FindMarkers(sn_atlas_processed,ident.1 = "DaN_0_PD_B5-6", ident.2 = "DaN_0_CTR")

deg_FC1andPadj1_numbers<-data.frame(celltype=c("Astrocyte_0","Astrocyte_1","Astrocyte_2","Astrocyte_3","DaN_0","DaN_1","DaN_2","DaN_3","GABA","Microglia_0","Microglia_1","Microglia_2","ODC_0","ODC_1","ODC_2","OPC","Tcells"),deg=c(length(astro0_d56[abs(astro0_d56$avg_log2FC)>=1 & astro0_d56$p_val_adj<=0.1,1]), length(astro1_d56[abs(astro1_d56$avg_log2FC)>=1 & astro1_d56$p_val_adj<=0.1,1]), 0, length(astro3_d56[abs(astro3_d56$avg_log2FC)>=1 & astro3_d56$p_val_adj<=0.1,1]), length(dan0_d56[abs(dan0_d56$avg_log2FC)>=1 & dan0_d56$p_val_adj<=0.1,1]), length(dan1_d56[abs(dan1_d56$avg_log2FC)>=1 & dan1_d56$p_val_adj<=0.1,1]), length(dan2_d56[abs(dan2_d56$avg_log2FC)>=1 & dan2_d56$p_val_adj<=0.1,1]), 0, length(gaba_d56[abs(gaba_d56$avg_log2FC)>=1 & gaba_d56$p_val_adj<=0.1,1]), length(micro0_d56[abs(micro0_d56$avg_log2FC)>=1 & micro0_d56$p_val_adj<=0.1,1]), length(micro1_d56[abs(micro1_d56$avg_log2FC)>=1 & micro1_d56$p_val_adj<=0.1,1]), length(micro2_d56[abs(micro2_d56$avg_log2FC)>=1 & micro2_d56$p_val_adj<=0.1,1]), length(odc0_d56[abs(odc0_d56$avg_log2FC)>=1 & odc0_d56$p_val_adj<=0.1,1]), length(odc1_d56[abs(odc1_d56$avg_log2FC)>=1 & odc1_d56$p_val_adj<=0.1,1]),  length(odc2_d56[abs(odc2_d56$avg_log2FC)>=1 & odc2_d56$p_val_adj<=0.1,1]),  length(opc_d56[abs(opc_d56$avg_log2FC)>=1 & opc_d56$p_val_adj<=0.1,1]),  length(tcell_d56[abs(tcell_d56$avg_log2FC)>=1 & tcell_d56$p_val_adj<=0.1,1])))


deg_FC1andPadj1_genes_UP<-list(Astrocyte_0=rownames(astro0_d56[astro0_d56$avg_log2FC>=1 & astro0_d56$p_val_adj<=0.05,]), Astrocyte_1=rownames(astro1_d56[astro1_d56$avg_log2FC>=1 & astro1_d56$p_val_adj<=0.05,]),
Astrocyte_3=rownames(astro3_d56[astro3_d56$avg_log2FC>=1 & astro3_d56$p_val_adj<=0.05,]),
DaN_0=rownames(dan0_d56[dan0_d56$avg_log2FC>=1 & dan0_d56$p_val_adj<=0.05,]),
DaN_1=rownames(dan1_d56[dan1_d56$avg_log2FC>=1 & dan1_d56$p_val_adj<=0.05,]),
DaN_2=rownames(dan2_d56[dan2_d56$avg_log2FC>=1 & dan2_d56$p_val_adj<=0.05,]),
GABA=rownames(gaba_d56[gaba_d56$avg_log2FC>=1 & gaba_d56$p_val_adj<=0.05,]),
Microglia_0=rownames(micro0_d56[micro0_d56$avg_log2FC>=1 & micro0_d56$p_val_adj<=0.05,]),
Microglia_1=rownames(micro1_d56[micro1_d56$avg_log2FC>=1 & micro1_d56$p_val_adj<=0.05,]),
Microglia_2=rownames(micro2_d56[micro2_d56$avg_log2FC>=1 & micro2_d56$p_val_adj<=0.05,]),
ODC_0=rownames(odc0_d56[odc0_d56$avg_log2FC>=1 & odc0_d56$p_val_adj<=0.05,]),
ODC_1=rownames(odc1_d56[odc1_d56$avg_log2FC>=1 & odc1_d56$p_val_adj<=0.05,]),
ODC_2=rownames(odc2_d56[odc2_d56$avg_log2FC>=1 & odc2_d56$p_val_adj<=0.05,]),
OPC=rownames(opc_d56[opc_d56$avg_log2FC>=1 & opc_d56$p_val_adj<=0.05,]),
Tcell=rownames(tcell_d56[tcell_d56$avg_log2FC>=1 & tcell_d56$p_val_adj<=0.05,])
)

deg_FC1andPadj1_genes_DOWN<-list(Astrocyte_0=rownames(astro0_d56[astro0_d56$avg_log2FC<= -1 & astro0_d56$p_val_adj<=0.05,]), Astrocyte_1=rownames(astro1_d56[astro1_d56$avg_log2FC<= -1 & astro1_d56$p_val_adj<=0.05,]),
Astrocyte_3=rownames(astro3_d56[astro3_d56$avg_log2FC<= -1 & astro3_d56$p_val_adj<=0.05,]),
DaN_0=rownames(dan0_d56[dan0_d56$avg_log2FC<= -1 & dan0_d56$p_val_adj<=0.05,]),
DaN_1=rownames(dan1_d56[dan1_d56$avg_log2FC<= -1 & dan1_d56$p_val_adj<=0.05,]),
DaN_2=rownames(dan2_d56[dan2_d56$avg_log2FC<= -1 & dan2_d56$p_val_adj<=0.05,]),
GABA=rownames(gaba_d56[gaba_d56$avg_log2FC<= -1 & gaba_d56$p_val_adj<=0.05,]),
Microglia_0=rownames(micro0_d56[micro0_d56$avg_log2FC<= -1 & micro0_d56$p_val_adj<=0.05,]),
Microglia_1=rownames(micro1_d56[micro1_d56$avg_log2FC<= -1 & micro1_d56$p_val_adj<=0.05,]),
Microglia_2=rownames(micro2_d56[micro2_d56$avg_log2FC<= -1 & micro2_d56$p_val_adj<=0.05,]),
ODC_0=rownames(odc0_d56[odc0_d56$avg_log2FC<= -1 & odc0_d56$p_val_adj<=0.05,]),
ODC_1=rownames(odc1_d56[odc1_d56$avg_log2FC<= -1 & odc1_d56$p_val_adj<=0.05,]),
ODC_2=rownames(odc2_d56[odc2_d56$avg_log2FC<= -1 & odc2_d56$p_val_adj<=0.05,]),
OPC=rownames(opc_d56[opc_d56$avg_log2FC<= -1 & opc_d56$p_val_adj<=0.05,]),
Tcell=rownames(tcell_d56[tcell_d56$avg_log2FC<= -1 & tcell_d56$p_val_adj<=0.05,])
)

pdf("deg_FC1andPadj1_ALL_numbers.pdf",10,3)
ggplot(deg_FC1andPadj1_ALL_numbers_melt,aes(x = cellType, y = variable,fill = value))+
    geom_tile() + scale_fill_gradient2(high = "red",mid = "red", low = "white",midpoint = 2500)+
    geom_tile() + theme(axis.text.x = element_text( color = "black",size = 10, angle = 45,hjust=1)) + geom_text(aes(label = round(value, 1))) 
dev.off()


jaccard <- function(a, b) {
    intersection = length(intersect(a, b))
    union = length(a) + length(b) - intersection
    return (intersection/union)
}

jac_ind<-data.frame(cellType=c("Astrocyte_0","Astrocyte_1","Astrocyte_3","DaN_0","DaN_1","DaN_2","GABA","Microglia_0","Microglia_1","Microglia_2","ODC_0","ODC_1","ODC_2","OPC","Tcells"),
JS_ILBD34_PD34=c(jaccard(deg_FC1andPadj05_ILBD34_genes$Astrocyte_0, deg_FC1andPadj05_PD34_genes$Astrocyte_0),
jaccard(deg_FC1andPadj05_ILBD34_genes$Astrocyte_1, deg_FC1andPadj05_PD34_genes$Astrocyte_1),
jaccard(deg_FC1andPadj05_ILBD34_genes$Astrocyte_3, deg_FC1andPadj05_PD34_genes$Astrocyte_3),
jaccard(deg_FC1andPadj05_ILBD34_genes$DaN_0, deg_FC1andPadj05_PD34_genes$DaN_0),
jaccard(deg_FC1andPadj05_ILBD34_genes$DaN_1, deg_FC1andPadj05_PD34_genes$DaN_1),
jaccard(deg_FC1andPadj05_ILBD34_genes$DaN_2, deg_FC1andPadj05_PD34_genes$DaN_2),
jaccard(deg_FC1andPadj05_ILBD34_genes$GABA, deg_FC1andPadj05_PD34_genes$GABA),
jaccard(deg_FC1andPadj05_ILBD34_genes$Microglia_0, deg_FC1andPadj05_PD34_genes$Microglia_0),
jaccard(deg_FC1andPadj05_ILBD34_genes$Microglia_1, deg_FC1andPadj05_PD34_genes$Microglia_1),
jaccard(deg_FC1andPadj05_ILBD34_genes$Microglia_2, deg_FC1andPadj05_PD34_genes$Microglia_2),
jaccard(deg_FC1andPadj05_ILBD34_genes$ODC_0, deg_FC1andPadj05_PD34_genes$ODC_0),
jaccard(deg_FC1andPadj05_ILBD34_genes$ODC_1, deg_FC1andPadj05_PD34_genes$ODC_1),
jaccard(deg_FC1andPadj05_ILBD34_genes$ODC_2, deg_FC1andPadj05_PD34_genes$ODC_2),
jaccard(deg_FC1andPadj05_ILBD34_genes$OPC, deg_FC1andPadj05_PD34_genes$OPC),
jaccard(deg_FC1andPadj05_ILBD34_genes$Tcell, deg_FC1andPadj05_PD34_genes$Tcell)),
JS_PD34_PD56=c(jaccard(deg_FC1andPadj05_PD56_genes$Astrocyte_0, deg_FC1andPadj05_PD34_genes$Astrocyte_0),
jaccard(deg_FC1andPadj05_PD56_genes$Astrocyte_1, deg_FC1andPadj05_PD34_genes$Astrocyte_1),
jaccard(deg_FC1andPadj05_PD56_genes$Astrocyte_3, deg_FC1andPadj05_PD34_genes$Astrocyte_3),
jaccard(deg_FC1andPadj05_PD56_genes$DaN_0, deg_FC1andPadj05_PD34_genes$DaN_0),
jaccard(deg_FC1andPadj05_PD56_genes$DaN_1, deg_FC1andPadj05_PD34_genes$DaN_1),
jaccard(deg_FC1andPadj05_PD56_genes$DaN_2, deg_FC1andPadj05_PD34_genes$DaN_2),
jaccard(deg_FC1andPadj05_PD56_genes$GABA, deg_FC1andPadj05_PD34_genes$GABA),
jaccard(deg_FC1andPadj05_PD56_genes$Microglia_0, deg_FC1andPadj05_PD34_genes$Microglia_0),
jaccard(deg_FC1andPadj05_PD56_genes$Microglia_1, deg_FC1andPadj05_PD34_genes$Microglia_1),
jaccard(deg_FC1andPadj05_PD56_genes$Microglia_2, deg_FC1andPadj05_PD34_genes$Microglia_2),
jaccard(deg_FC1andPadj05_PD56_genes$ODC_0, deg_FC1andPadj05_PD34_genes$ODC_0),
jaccard(deg_FC1andPadj05_PD56_genes$ODC_1, deg_FC1andPadj05_PD34_genes$ODC_1),
jaccard(deg_FC1andPadj05_PD56_genes$ODC_2, deg_FC1andPadj05_PD34_genes$ODC_2),
jaccard(deg_FC1andPadj05_PD56_genes$OPC, deg_FC1andPadj05_PD34_genes$OPC),
jaccard(deg_FC1andPadj05_PD56_genes$Tcell, deg_FC1andPadj05_PD34_genes$Tcell)))


```

### Pathway analysis (Figure 3)

```{r GOenrich_cellType, echo=TRUE}
library(topGO)

xx <- annFUN.org("BP", mapping = "org.Hs.eg.db", ID = "symbol")
allGenes <- unique(unlist(xx))

listGenesDE1<-as.character(rownames(df_w_test_level1[order(-df_w_test_level1$Astrocyte),])[1:15000])
listGenesDE1_fdr01<-as.character(deg_FC1andPadj1_genes_UP$Astrocyte_0)
universeGenes<-as.character(allGenes[allGenes%in%listGenesDE1])
geneList <- factor(as.integer(universeGenes %in% listGenesDE1_fdr01))
names(geneList)<-universeGenes
GOdata <- new("topGOdata",
              ontology = "BP",
              allGenes = geneList,
              nodeSize = 10,
              annot = annFUN.org, 
              mapping = "org.Hs.eg.db",
              ID = "symbol")
GOenrich<-runTest(GOdata, algorithm = "classic", statistic = "fisher")
resGO_deg_pd56_astrocyte0_up <- GenTable(GOdata, classic = GOenrich, ranksOf = "classic", topNodes = 100)


library(rrvgo)
simMatrix <- calculateSimMatrix(resGO_deg_pd56_astrocyte0_up$GO.ID,
orgdb="org.Hs.eg.db",
ont="BP",
method="Rel")
scores <- setNames(-log10(as.numeric(resGO_deg_pd56_astrocyte0_up$classic)), resGO_deg_pd56_astrocyte0_up$GO.ID)
reducedTerms <- reduceSimMatrix(simMatrix,
scores,
threshold=0.7,
orgdb="org.Hs.eg.db")
treemapPlot(reducedTerms)


```


### Pseudotime analysis for each cell type (Figure 3)

```{r pseudotime_cellType, echo=TRUE}

## DaNs:
library(monocle3)

genesALL_DaN<-read.delim("genesExp_DaNs_seuratIntegration_PC40res02",h=T)
genesALL_DaN<-as.matrix(genesALL_DaN)
metadataALL_DaN<-read.delim("metadata_DaNs_seuratIntegration_PC40res02",h=T)

gene_annotation<-data.frame(gene_short_name=rownames(genesALL_DaN),genes=rownames(genesALL_DaN))
rownames(gene_annotation)<-rownames(genesALL_DaN)

cds_dan <- new_cell_data_set(genesALL_DaN,
                             cell_metadata = metadataALL_DaN,
                             gene_metadata = gene_annotation)

cds_dan <- preprocess_cds(cds_dan, num_dim = 3)

cds_dan <- reduce_dimension(cds_dan)
cds_dan <- cluster_cells(cds_dan, resolution=1e-5)
cds_dan <- learn_graph(cds_dan)

plot_cells(cds_dan, color_cells_by = "Disease",label_branch_points = F,label_roots = F,label_leaves = F, cell_size=1)

cds_dan <- order_cells(cds_dan)
plot_cells(cds_dan, color_cells_by = "pseudotime",label_branch_points = F,label_roots = F,label_leaves = F, cell_size=1)


# density plot

pseud_dan<-pseudotime(cds_dan)
pseud_dan<-pseud_dan[is.finite(pseud_dan)]
metadataALL_DaN_filt<-metadataALL_DaN[rownames(metadataALL_DaN)%in%names(pseud_dan),]
metadataALL_DaN_filt$pseudotime<-pseud_dan

ggplot(metadataALL_DaN_filt, aes(x=pseudotime, fill=Disease)) +
    geom_density(alpha=.5) + theme_bw()

# alternatively:
pseudotime_DaNs_seuratIntegration_PC40res02<-read.delim("pseudotime_DaNs_seuratIntegration_PC40res02",h=F)
pseudotime_DaNs_seuratIntegration_PC40res02<-pseudotime_DaNs_seuratIntegration_PC40res02[match(meta_toplot$id,pseudotime_DaNs_seuratIntegration_PC40res02[,1]),]
meta_toplot$ps<-pseudotime_DaNs_seuratIntegration_PC40res02[,2]
ggplot(meta_toplot, aes(x=ps, fill=disease)) +
     geom_density(alpha=.5) + theme_bw()


# find DE genes along pseudotime (switchDE)


prot_cod_len<-read.delim("gene_length",h=T)
data<-as.matrix(GetAssayData(sn_atlas_processed_tmp, slot = "counts"))
prot_cod_m<-prot_cod_len[match(rownames(data),prot_cod_len[,1]),]
data <- sweep(data, 1, STATS = prot_cod_m$Gene_Length/1000, FUN = "/")
data[is.na(data)]<-0
data <- sweep(data, 2, STATS = colSums(data)/(10^6), FUN = "/")
SN_TPM_log_red<-log(data+1)
X_filtered <- SN_TPM_log_red[rowMeans(SN_TPM_log_red) > 0.1 & rowMeans(SN_TPM_log_red > 0) > 0.2,]
sde_dan <- switchde(X_filtered, pseud_dan)
sde_dan <-arrange(sde_dan,qval)

sde_dan<-sde_dan[sde_dan$qval<0.01 & abs(sde_dan$k)>0.03,]
ppi_net_sde_dan<-ppi_net[ppi_net[,1]%in%sde_dan$gene & ppi_net[,2]%in%sde_dan$gene ,]
ppi_net_sde_dan_genes<-unique(c(as.character(ppi_net_sde_dan[,1]),as.character(ppi_net_sde_dan[,2])))

net_1 <- graph_from_data_frame(d= ppi_net_sde_dan, vertices= ppi_net_sde_dan_genes, directed=F)
cl_net_1<-cluster_louvain(net_1)
memb_net_1<-as.matrix(membership(cl_net_1))

memb_net_1_df<-data.frame(id=rownames(memb_net_1),module=memb_net_1)
table(memb_net_1_df$module) #tmp<-read.delim("PPImodules_PPInetwork_switchDE_DAN_q01_k02")


# correlation with slingshot pseudotime:
# use sn_atlas_processed (with all cell types) and run Seurat classical pipeline, then subset only DaNs and run slingshot
sn_atlas_processed_tmp<-subset(sn_atlas_processed,subset=CellType=="DaN")
sce_dans <- as.SingleCellExperiment(sn_atlas_processed_tmp, assay = "RNA")
sce_dans <- slingshot(sce_dans, reducedDim = 'UMAP', clusterLabels = 'Disease')
meta_toplot<-data.frame(ps=sce_dans$slingPseudotime_1,disease=sce_dans$Disease,cellt=sce_dans$CellSubType)
ggplot(meta_toplot, aes(x=ps, fill=disease)) +
geom_density(alpha=.5) + theme_bw()

cor.test(sce_dans$slingPseudotime_1,pseudotime_DaNs_seuratIntegration_PC40res02[,2]) # corr = 0.61

# find DE genes along pseudotime (tradeseq)
# done for DaN_0 and DaN_1 subtypes across all disease condition

sn_atlas_processed_tmp<-subset(sn_atlas_processed,subset=CellSubType=="DaN_0" | CellSubType=="DaN_1")
sce_dans <- as.SingleCellExperiment(sn_atlas_processed_tmp, assay = "RNA")
sce_dans <- slingshot(sce_dans, reducedDim = 'UMAP', clusterLabels = 'Disease')
meta_toplot<-data.frame(ps=sce_dans$slingPseudotime_1,disease=sce_dans$Disease,cellt=sce_dans$CellSubType)
ggplot(meta_toplot, aes(x=ps, fill=disease)) +
     geom_density(alpha=.5) + theme_bw()

genes_dans<-as.matrix(GetAssayData(sn_atlas_processed_tmp, slot = "counts"))
sce_dans01_deg <- fitGAM(genes_dans,sds=SlingshotDataSet(sce_dans))
dans01_deg_tradeseq <- associationTest(sce_dans01_deg)
dans01_deg_tradeseqStartEnd <- startVsEndTest(sce_dans01_deg)
dim(dans01_deg_tradeseqStartEnd[dans01_deg_tradeseqStartEnd$pvalue<0.05 & dans01_deg_tradeseqStartEnd$logFClineage1>1,])
[1] 1371    4
dim(dans01_deg_tradeseqStartEnd[dans01_deg_tradeseqStartEnd$pvalue<0.05 & dans01_deg_tradeseqStartEnd$logFClineage1< -1,])
[1] 9381    4

ps_dans01<-cut(sce_dans$slingPseudotime_1,breaks = 3,labels = c("ps_int1","ps_int2","ps_int3"))
names(ps_dans01)<-colnames(sn_atlas_processed_tmp)
sn_atlas_processed_tmp$ps_interval<-ps_dans01
avg_exp_ps_dan01_down<-AverageExpression(object = sn_atlas_processed_tmp, group.by = c('ps_interval'),features = rownames(dans01_deg_tradeseqStartEnd[dans01_deg_tradeseqStartEnd$pvalue<0.05 & dans01_deg_tradeseqStartEnd$logFClineage1< -1,]))
avg_exp_ps_dan01_up<-AverageExpression(object = sn_atlas_processed_tmp, group.by = c('ps_interval'),features = rownames(dans01_deg_tradeseqStartEnd[dans01_deg_tradeseqStartEnd$pvalue<0.05 & dans01_deg_tradeseqStartEnd$logFClineage1> 1,]))
avg_exp_ps_dan01_all<-rbind(avg_exp_ps_dan01_up$RNA,avg_exp_ps_dan01_down$RNA)

pheatmap::pheatmap(log(avg_exp_ps_dan01_all+1), cluster_rows=F, cluster_cols=F,scale = "row",fontsize=2)

xx <- annFUN.org("BP", mapping = "org.Hs.eg.db", ID = "symbol")
allGenes <- unique(unlist(xx))
listGenesDE1<-as.character(rownames(genes_dans))
listGenesDE1_fdr01<-as.character(rownames(dans01_deg_tradeseqStartEnd[dans01_deg_tradeseqStartEnd$pvalue<0.05 & dans01_deg_tradeseqStartEnd$logFClineage1< -1,]))
universeGenes<-as.character(allGenes[allGenes%in%listGenesDE1])
geneList <- factor(as.integer(universeGenes %in% listGenesDE1_fdr01))
names(geneList)<-universeGenes
GOdata <- new("topGOdata",
              ontology = "BP",
              allGenes = geneList,
              nodeSize = 10,
              annot = annFUN.org, 
              mapping = "org.Hs.eg.db",
              ID = "symbol")
GOenrich<-runTest(GOdata, algorithm = "classic", statistic = "fisher")
resGO_dans01_deg_tradeseqStartEnd_down <- GenTable(GOdata, classic = GOenrich, ranksOf = "classic", topNodes = 100)


df_go<-resGO_dans01_deg_tradeseqStartEnd_down[c(3,6,10,21,24,26,30,41,50),c(2,6)]
df_go$Term<-factor(df_go$Term,levels = as.character(df_go$Term))
ggplot(data=df_go, aes(x=Term, y=-log(as.numeric(classic)))) +
     geom_bar(stat="identity", fill="skyblue")+
     theme_minimal() + coord_flip()




## ODCs:
sn_atlas_processed_tmp<-subset(sn_atlas_processed,subset=CellType=="ODC")
sn_atlas_processed_tmp <- NormalizeData(sn_atlas_processed_tmp)
sn_atlas_processed_tmp <- FindVariableFeatures(sn_atlas_processed_tmp, selection.method = "vst", nfeatures = 2000)
sn_atlas_processed_tmp <- ScaleData(sn_atlas_processed_tmp)
sn_atlas_processed_tmp <- RunPCA(sn_atlas_processed_tmp, features = VariableFeatures(object = sn_atlas_processed_tmp))
sn_atlas_processed_tmp <- FindNeighbors(sn_atlas_processed_tmp, dims = 1:30)
sn_atlas_processed_tmp <- FindClusters(sn_atlas_processed_tmp, resolution = 0.6)
sn_atlas_processed_tmp <- RunUMAP(sn_atlas_processed_tmp, dims = 1:30)
sce_odc <- as.SingleCellExperiment(sn_atlas_processed_tmp, assay = "RNA")
sce_odc <- slingshot(sce_odc, reducedDim = 'UMAP', clusterLabels = 'Disease')


meta_toplot_odc<-data.frame(ps=sce_odc$slingPseudotime_1,disease=sce_odc$Disease,cellt=sce_odc$CellSubType)
meta_toplot_odc<-meta_toplot_odc[meta_toplot_odc$cellt=="ODC_2",]
ggplot(meta_toplot_odc, aes(x=ps, fill=disease)) +
    geom_density(alpha=.5) + theme_bw()


# characterisation of the genes changing expression along pseudotime

genes_odc2<-as.matrix(GetAssayData(sn_atlas_processed_tmp, slot = "counts"))
sce_odc2_deg <- fitGAM(genes_odc2,sds=SlingshotDataSet(sce_odc2))
odc2_deg_tradeseq <- associationTest(sce_odc2_deg)
odc2_deg_tradeseqStartEnd <- startVsEndTest(sce_odc2_deg)

library(clusterExperiment)
nPointsClus <- 20
clusPat <- clusterExpressionPatterns(sce_odc2_deg, nPoints = nPointsClus,
                                     genes = rownames(odc2_deg_tradeseqStartEnd[odc2_deg_tradeseqStartEnd$pvalue<0.05,]))

ppi_net_short_odc2sce_upreg<-ppi_net_short[ppi_net_short[,1]%in%rownames(odc2_deg_tradeseqStartEnd[odc2_deg_tradeseqStartEnd$pvalue<0.05 & odc2_deg_tradeseqStartEnd$logFClineage1>1,]) & ppi_net_short[,2]%in%rownames(odc2_deg_tradeseqStartEnd[odc2_deg_tradeseqStartEnd$pvalue<0.05 & odc2_deg_tradeseqStartEnd$logFClineage1>1,]),]
ppi_net_short_odc2sce_upreg_genes<-unique(c(as.character(ppi_net_short_odc2sce_upreg[,1]),as.character(ppi_net_short_odc2sce_upreg[,2])))
net_opc <- graph_from_data_frame(d= ppi_net_short_odc2sce_upreg, vertices= ppi_net_short_odc2sce_upreg_genes, directed=F)
cl_net<-cluster_louvain(net_opc)
memb_net<-as.matrix(membership(cl_net))
memb_net_odc2sce_upreg_ppishort <-data.frame(id=rownames(memb_net),module=memb_net)
table(memb_net_odc2sce_upreg_ppishort$module)

xx <- annFUN.org("BP", mapping = "org.Hs.eg.db", ID = "symbol")
 allGenes <- unique(unlist(xx))
 listGenesDE1<-as.character(rownames(odc2_deg_tradeseqStartEnd))
 listGenesDE1_fdr01<-as.character(memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==1,1])
 universeGenes<-as.character(allGenes[allGenes%in%listGenesDE1])
 geneList <- factor(as.integer(universeGenes %in% listGenesDE1_fdr01))
 names(geneList)<-universeGenes
 GOdata <- new("topGOdata",
 ontology = "BP",
 allGenes = geneList,
 nodeSize = 10,
 annot = annFUN.org,
 mapping = "org.Hs.eg.db",
 ID = "symbol")
GOenrich<-runTest(GOdata, algorithm = "classic", statistic = "fisher")
resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M1 <- GenTable(GOdata, classic = GOenrich, ranksOf = "classic", topNodes = 100)

tmp_up_m1<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==1,1],]
tmp_up_m2<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==2,1],]
tmp_up_m3<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==3,1],]
tmp_up_m4<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==4,1],]
tmp_up_m5<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==5,1],]
tmp_up_m6<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==6,1],]
tmp_up_m7<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==7,1],]
tmp_up_m8<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==8,1],]
tmp_up_m9<-clusPat$yhatScaled[rownames(clusPat$yhatScaled)%in%memb_net_odc2sce_upreg_ppishort[memb_net_odc2sce_upreg_ppishort$module==9,1],]

tmp_up_m1<-colMeans(tmp_up_m1)
tmp_up_m2<-colMeans(tmp_up_m2)
tmp_up_m3<-colMeans(tmp_up_m3)
tmp_up_m4<-colMeans(tmp_up_m4)
tmp_up_m5<-colMeans(tmp_up_m5)
tmp_up_m6<-colMeans(tmp_up_m6)
tmp_up_m7<-colMeans(tmp_up_m7)
tmp_up_m8<-colMeans(tmp_up_m8)
tmp_up_m9<-colMeans(tmp_up_m9)

tmp<-rbind(tmp_up_m9,tmp_up_m8,tmp_up_m7,tmp_up_m6,tmp_up_m5,tmp_up_m4,tmp_up_m3,tmp_up_m2,tmp_up_m1) 
rownames(tmp)<-c("M9","M8","M7","M6","M5","M4","M3","M2","M1")
pheatmap::pheatmap(tmp, cluster_rows=F, cluster_cols=F,scale = "row",fontsize=7,border_color = NA)

tmp<-rbind(resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M1[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M2[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M3[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M4[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M5[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M6[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M7[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M8[1:3,c(2,6)],resGO_sce_odc2_tradeseqStartEnd_PPImod_upreg_M9[1:3,c(2,6)])
tmp$Mod<-c(rep("M1",3),rep("M2",3),rep("M1",3),rep("M2",3),rep("M1",3),rep("M2",3),rep("M1",3),rep("M2",3),rep("M1",3))

tmp$classic<-as.numeric(tmp$classic)
tmp$Term<-factor(tmp$Term,levels = as.character(tmp$Term))
ggplot(tmp, aes(x = -log(classic), y = Term,fill=Mod)) + geom_bar(stat = "identity") + theme_minimal() + scale_fill_manual(values = c("blue","skyblue")) + theme(axis.text.y  = element_text(hjust=0,face = "bold"))

```


### Isoform Analysis (Figure 3)

```{r isoform, echo=TRUE}
library(fishpond)
library(scran)
library(biomaRt)
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "https://useast.ensembl.org")
library(tximeta)

coldata<-read.delim("coldata_ALL_CTRandPD56_modified_nfs",h=T,stringsAsFactors=F)

suppressPackageStartupMessages(library(SummarizedExperiment))
y <- tximeta(coldata,dropInfReps=T)

# diff transcript express

y <- labelKeep(y, minCount = 3, minN = 10)
y <- y[mcols(y)$keep,]
set.seed(1)
assays(y) <- lapply(assays(y), as.matrix) # make dense matrices
y <- scaleInfReps(y, lengthCorrect=FALSE, sfFun= computeSumFactors)
y <- swish(y, x="condition",  quiet=TRUE)

# diff isoform usage

iso <- isoformProportions(y)
iso <- swish(iso, x="condition",nperms=64)

df<-mcols(iso)[,c("log2FC","qvalue","gene","tx_id")]

load("Transcript_expression_allCelltype_CTRandPD56.RData") # to get all data and get number of DTUs

# OPCs to ODCs differentiation using isoform data

load("pseudotime_isoforms_OPC_ODCs_CTRandPD56.RData")

isoform_OPC_ODC2_CTRandPD56<-as.matrix(assays(y)[[2]])
meta_isoform_OPC_ODC2_CTRandPD56<-data.frame(celltype=y$cellSubType,disease=y$condition,cell=y$Barcode,sample=y$names)
meta_isoform_OPC_ODC2_CTRandPD56$names<-paste(meta_isoform_OPC_ODC2_CTRandPD56$cell,meta_isoform_OPC_ODC2_CTRandPD56$sample,sep="_")
colnames(isoform_OPC_ODC2_CTRandPD56)<-meta_isoform_OPC_ODC2_CTRandPD56$names
rownames(meta_isoform_OPC_ODC2_CTRandPD56)<-meta_isoform_OPC_ODC2_CTRandPD56$names

chip_iso_opc_odc2_ctrpd <- CreateSeuratObject(counts = isoform_OPC_ODC2_CTRandPD56, project = "isoform", min.cells = 0, min.features = 0,meta.data= meta_isoform_OPC_ODC2_CTRandPD56)

# done on controls and pd samples separately
chip_iso_opc_odc2_ctr <- subset(chip_iso_opc_odc2_ctrpd,subset=disease=="CTR")
chip_iso_opc_odc2_ctr <- NormalizeData(chip_iso_opc_odc2_ctr)
chip_iso_opc_odc2_ctr <- FindVariableFeatures(chip_iso_opc_odc2_ctr, selection.method = "vst", nfeatures = 2000)

chip_iso_opc_odc2_ctr <- ScaleData(chip_iso_opc_odc2_ctr)
chip_iso_opc_odc2_ctr <- RunPCA(chip_iso_opc_odc2_ctr, features = VariableFeatures(object = chip_iso_opc_odc2_ctrpd))
chip_iso_opc_odc2_ctr <- FindNeighbors(chip_iso_opc_odc2_ctr, dims = 1:10)
chip_iso_opc_odc2_ctr <- FindClusters(chip_iso_opc_odc2_ctr, resolution = 0.6)
chip_iso_opc_odc2_ctr <- RunUMAP(chip_iso_opc_odc2_ctr, dims = 1:10)

sce_iso_opc_odc2_ctr <- as.SingleCellExperiment(chip_iso_opc_odc2_ctr, assay = "RNA")
sce_iso_opc_odc2_ctr <- slingshot(sce_iso_opc_odc2_ctr, reducedDim = 'UMAP', clusterLabels = 'seurat_clusters')
meta_isoform_OPC_ODC2_CTR<-meta_isoform_OPC_ODC2_CTRandPD56[meta_isoform_OPC_ODC2_CTRandPD56$disease=="CTR",]
meta_isoform_OPC_ODC2_CTR$ps<-sce_iso_opc_odc2_ctr$slingPseudotime_1

ggplot(meta_isoform_OPC_ODC2_CTR, aes(x=ps, fill=celltype)) +
    geom_density(alpha=.5) + theme_bw()

isoform_OPC_ODC2_CTR<-as.matrix(GetAssayData(chip_iso_opc_odc2_ctr, slot = "counts"))
isoform_OPC_ODC2_CTR<-log(isoform_OPC_ODC2_CTR+1)
sde_iso_opc_odc2_ctr <- switchde(isoform_OPC_ODC2_CTR, sce_iso_opc_odc2_ctr$slingPseudotime_1)

pd_risk_opc_isoform<-read.table("pd_50k_gene_and_iso_opc.gsa.sets.genes.out",h=T)
pd_risk_odc2_isoform<-read.table("pd_50k_gene2000_and_iso_odc2_only.gsa.sets.genes.out",h=T)

# test enrichment of pd risk isoforms in OPCs and ODC_2 along differentiation trajectory:


meta_isoform_OPC_ODC_CTR_tmp<-meta_isoform_OPC_ODC2_CTRandPD56[meta_isoform_OPC_ODC2_CTRandPD56$celltype=="OPC",]
meta_isoform_OPC_ODC_CTR_tmp <-meta_isoform_OPC_ODC_CTR_tmp[order(meta_isoform_OPC_ODC_CTR_tmp$ps),]
meta_isoform_OPC_ODC_CTR_tmp<-meta_isoform_OPC_ODC_CTR_tmp[meta_isoform_OPC_ODC_CTR_tmp$ps>10,] # opcs subset
meta_isoform_OPC_ODC_CTR_tmp$set<-ifelse(meta_isoform_OPC_ODC_CTR_tmp$ps<15,"late","early")
chip_iso_opc_odc_ctr_tmp<-chip_iso_opc_odc2_ctr[,colnames(chip_iso_opc_odc2_ctr)%in%meta_isoform_OPC_ODC_CTR_tmp$names]
chip_iso_opc_odc_ctr_tmp$set<-meta_isoform_OPC_ODC_CTR_tmp$set

Idents(chip_iso_opc_odc_ctr_tmp)<-chip_iso_opc_odc_ctr_tmp$set
opc_latevsearly_ctr<-FindMarkers(chip_iso_opc_odc_ctr_tmp,ident.1 = "late", ident.2 = "early")

dim(opc_latevsearly_ctr[opc_latevsearly_ctr$p_val_adj<=0.05,])
tmp<-prot_cod[match(rownames(opc_latevsearly_ctr),prot_cod[,2]),]
opc_latevsearly_ctr$id<-tmp$Gene.name

# GO enrichment for early and late genes along differentiation trajectory:
xx <- annFUN.org("BP", mapping = "org.Hs.eg.db", ID = "symbol")
allGenes <- unique(unlist(xx))
listGenesDE1<-as.character(rownames(Wtest_counts_L1[order(Wtest_counts_L1$OPC),])[1:5000])
listGenesDE1_fdr01<-as.character(opc_latevsearly_ctr[opc_latevsearly_ctr$p_val<=0.05 & opc_latevsearly_ctr$avg_log2FC>0,6])
universeGenes<-as.character(allGenes[allGenes%in%listGenesDE1])
geneList <- factor(as.integer(universeGenes %in% listGenesDE1_fdr01))
names(geneList)<-universeGenes
GOdata <- new("topGOdata",
              ontology = "BP",
              allGenes = geneList,
              nodeSize = 10,
              annot = annFUN.org, 
              mapping = "org.Hs.eg.db",
              ID = "symbol")
GOenrich<-runTest(GOdata, algorithm = "classic", statistic = "fisher")
resGO_opc_latevsearly_ctr_up <- GenTable(GOdata, classic = GOenrich, ranksOf = "classic", topNodes = 100)

```


### Cell-cell communication analysis (Figure 4)

```{r cellcomm, echo=TRUE}

library(CellChat)
library(patchwork)

sn_atlas_ctr<-subset(sn_atlas_processed,subset=Disease=="CTR")
sn_atlas_pd<-subset(sn_atlas_processed,subset=Disease=="PD_B5-6")

Idents(sn_atlas_ctr)<-sn_atlas_ctr$CellSubType
sn_atlas_ctr_downsample<-subset(x = sn_atlas_ctr, downsample = 150)
Idents(sn_atlas_pd)<-sn_atlas_pd$CellSubType
sn_atlas_pd_downsample<-subset(x = sn_atlas_pd, downsample = 150)

data.input <- GetAssayData(sn_atlas_ctr_downsample, assay = "RNA", slot = "data") 
labels <-sn_atlas_ctr_downsample$CellSubType
names(labels)<-colnames(data.input)
meta <- data.frame(group = labels, row.names = names(labels))
cellchat.CTR <- createCellChat(object = data.input, meta = meta, group.by = "group")

data.input <- GetAssayData(sn_atlas_pd_downsample, assay = "RNA", slot = "data") 
labels <-sn_atlas_pd_downsample$CellSubType
names(labels)<-colnames(data.input)
meta <- data.frame(group = labels, row.names = names(labels))
cellchat.PD <- createCellChat(object = data.input, meta = meta, group.by = "group")


# to run on each dataset separately

CellChatDB <- CellChatDB.human
CellChatDB.use <- CellChatDB
cellchat.CTR@DB <- CellChatDB.use
cellchat.PD@DB <- CellChatDB.use

cellchat.CTR <- subsetData(cellchat.CTR)
cellchat.PD <- subsetData(cellchat.PD)
cellchat.CTR <- identifyOverExpressedGenes(cellchat.CTR)
cellchat.PD <- identifyOverExpressedGenes(cellchat.PD)


cellchat.CTR <- identifyOverExpressedInteractions(cellchat.CTR)
cellchat.PD <- identifyOverExpressedInteractions(cellchat.PD)


#cellchat.CTR <- projectData(cellchat.CTR, PPI.human)
#cellchat.PD <- projectData(cellchat.PD, PPI.human)

cellchat.CTR <- computeCommunProb(cellchat.CTR,raw.use=TRUE,population.size = FALSE)
cellchat.PD <- computeCommunProb(cellchat.PD,raw.use=TRUE,population.size = FALSE)

cellchat.CTR <- filterCommunication(cellchat.CTR, min.cells = 10)
df.net.CTR <- subsetCommunication(cellchat.CTR)
df.netP.CTR <- subsetCommunication(cellchat.CTR,slot.name = "netP")
write.table(df.net.CTR,"SNatlas_communicationNetwork_results_CTR_downsampled",quote=F,sep="\t",row.names=F)
write.table(df.netP.CTR,"SNatlas_communicationNetwork_signalingPath_results_CTR_downsampled",quote=F,sep="\t",row.names=F)
cellchat.CTR <- computeCommunProbPathway(cellchat.CTR)
cellchat.CTR <- aggregateNet(cellchat.CTR)

cellchat.PD <- filterCommunication(cellchat.PD, min.cells = 10)
df.net.PD <- subsetCommunication(cellchat.PD)
df.netP.PD <- subsetCommunication(cellchat.PD,slot.name = "netP")
write.table(df.net.PD,"SNatlas_communicationNetwork_results_PD_downsampled",quote=F,sep="\t",row.names=F)
write.table(df.netP.PD,"SNatlas_communicationNetwork_signalingPath_results_PD_downsampled",quote=F,sep="\t",row.names=F)
cellchat.PD <- computeCommunProbPathway(cellchat.PD)
cellchat.PD <- aggregateNet(cellchat.PD)


cellchat.CTR <- netAnalysis_computeCentrality(cellchat.CTR, slot.name = "netP")
netAnalysis_signalingRole_heatmap(cellchat.CTR, pattern = "outgoing",font.size=4,height=20,width=6)
netAnalysis_signalingRole_heatmap(cellchat.CTR, pattern = "incoming",font.size=4,height=20,width=6)

cellchat.PD <- netAnalysis_computeCentrality(cellchat.PD, slot.name = "netP")
netAnalysis_signalingRole_heatmap(cellchat.PD, pattern = "outgoing",font.size=4,height=20,width=6)
netAnalysis_signalingRole_heatmap(cellchat.PD, pattern = "incoming",font.size=4,height=20,width=6)

# to merge

group.new = levels(cellchat.CTR@idents)
cellchat.PD <- liftCellChat(cellchat.PD, group.new)
object.list <- list(CTR = cellchat.CTR, PD = cellchat.PD)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")
gg1 + gg2

gg1 <- netVisual_heatmap(cellchat)
gg2 <- netVisual_heatmap(cellchat, measure = "weight")
gg1 + gg2

df.net.CTR <- subsetCommunication(cellchat.CTR)
df.net.PD <- subsetCommunication(cellchat.PD)


# test enrichment in PD genetic risk associated genes in ligand/receptor interactions betwen any cell type

cellcomm<-df.net.CTR
Wtest_DaNs<-read.delim("Wtest_DaNs",h=T)
dan3<-Wtest_DaNs[order(-Wtest_DaNs$DaN_3),] # to obtain DaN_3 specific genes
pd_risk_odc2<-read.delim("MAGMA_pd_risk_odcs2",h=T) # to obtain pd risk associated genes to ODC_2

# Here it is done for ODC_2 and DaN_3 interactions. It has to be repeated for any pair:
p_test<-c()
test_interact1<-cellcomm[(cellcomm$target=="DaN_3" & cellcomm$source=="ODC_2" & cellcomm$receptor%in%rownames(dan3)[1:1000])  ,]
test_interact2<-cellcomm[(cellcomm$source=="DaN_3" & cellcomm$target=="ODC_2" & cellcomm$ligand%in%rownames(dan3)[1:1000])  ,]
test_interact3<-cellcomm[(cellcomm$target=="DaN_3" & cellcomm$source=="ODC_2")  ,]
test_interact4<-cellcomm[(cellcomm$source=="DaN_3" & cellcomm$target=="ODC_2")  ,]


tmp4<-test_interact1[test_interact1$ligand%in%pd_risk_odc2$ID[1:200] ,]
tmp5<-test_interact2[test_interact2$receptor%in%pd_risk_odc2$ID[1:200] ,]
tmp6<-unique(c(tmp4$ligand,tmp5$receptor))
if(length(tmp6)==0){
    p_test<-c(p_test,1)
}else{
    p=0
    for(i in 1:10000){
        tmp_sample1<-sample(pd_risk_odc2$ID[201:1800],200)
	tmp_interact1<-sample(test_interact3$ligand,length(unique(test_interact1$ligand)))
	tmp_interact2<-sample(test_interact4$receptor,length(unique(test_interact2$receptor)))
        
        tmp2<-tmp_interact1[ tmp_interact1%in%tmp_sample1  ]
        tmp3<-tmp_interact2[ tmp_interact2%in%tmp_sample1  ]
        tmp7<-unique(c(tmp2,tmp3))
        if(length(tmp7)>length(tmp6)){
            p=p+1}
    }
    p_test<-c(p_test,p/10000)
}
p_test

```


### Cell state transition DaN_1 to DaN_3 (Figure 5)

```{r state_transition, echo=TRUE}

sn_atlas_processed_tmp<-subset(sn_atlas_processed,subset=CellSubType=="DaN_1" | CellSubType=="DaN_3")
sn_atlas_processed_tmp<-subset(sn_atlas_processed,subset=Disease=="CTR")
sce_dans <- as.SingleCellExperiment(sn_atlas_processed_tmp, assay = "RNA")
sce_dans <- slingshot(sce_dans, reducedDim = 'UMAP', clusterLabels = 'CellSubType')
meta_toplot<-data.frame(ps=sce_dans$slingPseudotime_1,cellt=sce_dans$CellSubType)
ggplot(meta_toplot, aes(x=ps, fill=cellt)) +
geom_density(alpha=.5) + theme_bw()

prot_cod_len<-read.delim("gene_length",h=T)
data<-as.matrix(GetAssayData(sn_atlas_processed_tmp, slot = "counts"))
prot_cod_m<-prot_cod_len[match(rownames(data),prot_cod_len[,1]),]
data <- sweep(data, 1, STATS = prot_cod_m$Gene_Length/1000, FUN = "/")
data[is.na(data)]<-0
data <- sweep(data, 2, STATS = colSums(data)/(10^6), FUN = "/")
SN_TPM_log_red<-log(data+1)
X_filtered <- SN_TPM_log_red[rowMeans(SN_TPM_log_red) > 0.1 & rowMeans(SN_TPM_log_red > 0) > 0.2,]
sde_dan <- switchde(X_filtered, sce_dans$slingPseudotime_1)
sde_dan <-arrange(sde_dan,qval)

sde_dan<-sde_dan[sde_dan$qval<0.01 & abs(sde_dan$k)>0.03,]


```


### Coexpression network analysis

```{r coexpr, echo=TRUE}

data_dan3<-as.matrix(GetAssayData(sn_atlas_DAN3, slot = "counts"))
results.dan3=compute.network(expr.data = data_dan3,gene.names = rownames(data_dan3),clustering = "direct")

m<-data.frame(results.dan3$correlations)
coexp_dan3<-data.frame(row=rownames(m)[row(m)[upper.tri(m)]], 
                       col=colnames(m)[col(m)[upper.tri(m)]], 
                       corr=m[upper.tri(m)])


coexp_dan3_08<-coexp_dan3[abs(coexp_dan3$corr)>0.8,]

coexp_dan3_08_genes<-unique(c(as.character(coexp_dan3_08[,1]),as.character(coexp_dan3_08[,2])))
net_dan3 <- graph_from_data_frame(d= coexp_dan3_08, vertices= coexp_dan3_08_genes, directed=F)
cl_net<-cluster_louvain(net_dan3)
memb_net_dan3<-as.matrix(membership(cl_net))
memb_net_dan3_df<-data.frame(id=rownames(memb_net_dan3),module=memb_net_dan3)

```














